### ISSUE TITLE
Cant click on objects behind players

### ISSUE THREAD

**Issue by Dodg-e** [jasonrohrer/OneLife#79]
For example filling a bowl with berries when someone stands in front of the bowl

**Comment by jasonrohrer** [comment]
This is a deeper issue that is much harder to fix.

You do sometimes need to click on players, and sometimes on objects.

If a player is standing in front of a small object, then yes, it seems like you should be able to click "through" them to get the hidden object.  But if they are standing in front of a large object, one that covers their entire body from behind, then allowing you to click through them makes it impossible to click on them (anywhere you click on them, you click the big object behind them).

There is no obvious heuristic for determining which case applies.  Which objects are "big enough" that you should not be able to click through a player to get to them?

You could imagine a perfectly-shaped object, like an oval-shapped vase, that exactly matches the player's body shape.  It would certainly seem too big.... or maybe too small?

Going to leave this one hanging for now.  Hopefully a clever idea to solve this problem will arise.

**Comment by Dodg-e** [comment]
If the number of pixels of the sprite is below a certain number (small object) then allow transparency else dont allow transparency (big object) 

Could that work?

**Comment by versality** [comment]
I think making a specific modifier hotkey can help and make it a reliable experience.
During interaction, if certain key is pressed, it will always target player regardless of his sprite overlap.

If not, just use whatever logic is currently in use.

From what I remember Don't starve had a similar problem bit with the attack action.
If there's an item behind a creature, if you click it, it would attack the creature instead.
Using key modifier would prevent that (or vice versa).

**Comment by swallis21** [comment]
Key modifiers go against the two-button design of the game. Of course you're going to run into issues like this, though, when your designed mechanics are so deliberately simple. And you've already committed to right-clicking on players being reserved for killing them (for good reason...) so you're in a bit of a corner. Fundamentally change your mechanics, or let players die because someone was standing in the way of something they needed? I saw a suggestion on the forums that the middle mouse button be used for this. It seems like the best option, though still not without flaws.

FYI this is why games from this perspective squish everything into one tile.

**Comment by jasonrohrer** [comment]
What do you mean squish everything into one tile?

Oh, you mean, so that the player cannot overlap the stuff in the tile to the north of them?

**Comment by swallis21** [comment]
Yeah, all the graphics for something on a tile don't exceed the boundaries of the tile. Not just players, but trees and walls and such too.

**Comment by JustinLove** [comment]
I've been experimenting with preferring objects during screen-lock. It makes sense because you often have the screen locked during time critical work like smiting where people could get in the way.  I don't consider quite good enough because eating is also often a time critical operation where I naturally screen lock, and you have to aim at the lower body with this change. I thought about an exception for food, and then saw someone get blocked while trying to fill a berry bowl.

https://github.com/JustinLove/onelife-client-patches/blob/master/click-through-heads

**Comment by psilove** [comment]
I think I wrote a Reddit entry some month ago where I wished for a [Shift]+Click Option to always target the object behind a player or tree (!). Could be an easy solution to this. Clicking [Shift] could also make some objects around you half-transparent, so you can see what's behind a tree for example? In the real world you should be able to. :)

**Comment by Leonard2** [comment]
I see people have already suggested a key modifier.
I think it's the only way to address this issue.

Personally, I have found myself wishing for a modifier that would always interact with an object regardless of where I click on the given tile.
I recently found out right clicking on a tile with stackable items like flat stones/piles of iron/steel does this.
It has dramatically sped up this part of the game for me.
It doesn't work with stacks that can be carried (like plates or bowls) unfortunately.

I suggest having a modifier that works with both left and right click (since they both do different things depending on context) to make the action always work regardless of where you click on the given tile.
That way you kill two birds with one stone: people standing in front of items, and items that are generally very small and a huge pain to click on like keys or wet clay plates or single steel ingots while smithing.

I have been thinking of implementing this on my custom client recently.
I also really love psilove's idea of making such a modifier make tall objects that are being hovered over become translucent.
That would immensely help against griefers, because the highlight is often not enough at all for small objects like keys, though that's for another issue completely.
